File: csharp/strain/Strain.cs
using System;
using System.Collections.Generic;

public static class Strain
{
  public static IEnumerable<T> Keep<T>(this IEnumerable<T> collection, Func<T, bool> predicate)
  {
    return Filter(collection, predicate);
  }

  public static IEnumerable<T> Discard<T>(this IEnumerable<T> collection, Func<T, bool> predicate)
  {
    return Filter(collection, x => !predicate(x));
  }

  private static IEnumerable<T> Filter<T>(IEnumerable<T> source, Func<T, bool> predicate)
  {
    List<T> result = new List<T>();
    foreach (T item in source)
    {
      if (predicate(item))
      {
        result.Add(item);
      }
    }
    return result;
  }
}

File: javascript/strain/strain.js
var strain = {
  keep: function(collection, predicate) {
    return filter(collection, predicate);
  },
  discard: function(collection, predicate) {
    return filter(collection, x => !predicate(x));
  }
};

function filter(collection, predicate) {
  var output = [];
  for (var i = 0; i < collection.length; i++) {
    if (predicate(collection[i])) output.push(collection[i]);
  }
  return output;
};

module.exports = strain;
